std/typecons.d

```diff
@@ -9035,3 +9035,220 @@ unittest
+}
+
+@safe unittest
+{
+    import std.range : iota;
+    import core.bitop : bsr;
+    static struct S(size_t a)
+    {
+        bool b;
+        AlignedStorage!(1, 2^^a) s;
+    }
+    // test all alignments up to and including platformAlignment
+    foreach (a; aliasSeqOf!(iota(bsr(platformAlignment)+1)))
+    {
+        static assert(S!a.s.offsetof == 2^^a);
+    }
+}
+
+private
+{
+    static if (is(cent))
+        alias CentType = cent;
+    else
+        alias CentType = AliasSeq!();
+}
+
+@system unittest
+{
+    foreach (T; AliasSeq!(bool,byte,char,short,wchar,int,
+                          dchar,long,float,double,real,CentType))
+    {{
+        AlignedStorage!(T.sizeof, T.alignof) s;
+        *(cast(T*) s.ptr) = cast(T) 42;
+        T a = cast(T) 42;
+        auto raw = (cast(void*) &a)[0 .. T.sizeof];
+        assert(s[0 .. $] == raw[0 .. $]);
+        assert(s[][0 .. $] == raw[0 .. $]);
+    }}
+}
+
+/**
+Provides untyped storage suitable for storing an instance of type `T`.
+For classes, the size and alignment of storage are the size and alignment
+of class instance representation, not those of class reference.
+*/
+template InstanceStorage(T)
+if (!is(T == interface))
+{
+    static if (is(T == class))
+    {
+        alias InstanceStorage = AlignedStorage!(__traits(classInstanceSize, T),
+            classInstanceAlignment!T);
+    }
+    else
+    {
+        alias InstanceStorage = AlignedStorage!(T.sizeof, T.alignof);
+    }
+}
+
+///
+@safe unittest
+{
+    class SafeClass
+    {
+        int x;
+        @safe this(int x) { this.x = x; }
+    }
+
+    InstanceStorage!SafeClass buf;
+    auto support = (() @trusted => cast(SafeClass) (buf.ptr))();
+    import std.conv : emplace;
+    auto safeClass = emplace!SafeClass(support, 5);
+    scope (exit) () @trusted { destroy(safeClass); } ();
+    assert(safeClass.x == 5);
+}
+
+@safe unittest
+{
+    foreach (T; AliasSeq!(bool,byte,char,short,wchar,int,
+                          dchar,long,float,double,real,CentType))
+    {
+        static assert(InstanceStorage!T.sizeof == T.sizeof);
+        static assert(InstanceStorage!T.alignof == T.alignof);
+    }
+}
+
+@safe unittest
+{
+    static assert(InstanceStorage!Object.sizeof == __traits(classInstanceSize, Object));
+    static assert(InstanceStorage!Object.alignof == classInstanceAlignment!Object);
+
+    class Arbitrary
+    {
+        bool b;
+        float[3] v;
+        char[17] c;
+    }
+
+    static assert(InstanceStorage!Arbitrary.sizeof == __traits(classInstanceSize, Arbitrary));
+    static assert(InstanceStorage!Arbitrary.alignof == classInstanceAlignment!Arbitrary);
+}
+
+@safe unittest
+{
+    struct Arbitrary
+    {
+        bool b;
+        float[3] v;
+        char[17] c;
+    }
+
+    static assert(InstanceStorage!Arbitrary.sizeof == Arbitrary.sizeof);
+    static assert(InstanceStorage!Arbitrary.alignof == Arbitrary.alignof);
+}
+
+private size_t platformAlignment() @safe @nogc nothrow pure
```

Comment:
I think we have something like this in std.experimental.allocator.

Reply:
Yes. What would you suggest? Make this `package` and import in `std.experimental.allocator`? Or perhaps even `public`? User code could benefit from that information.

Reply:
Is it public in `std.experimental.allocator`? If so `public`, if not `package` and unify the definitions.

Reply:
It is `public`.